### PR TITLE
Add `Piece#valid_vector?`

### PR DESCRIPTION
**Pull request by swcraig**
This method calls a piece's private `#movement_vectors` method. This way, a
piece does not need to know either of its source or destination cells, just the
desired movement vector.  This is the method that the future `Move` object will
use to ask the moving piece whether it is able to move along the vector given.

### CODE REVIEW

File: spec/piece_specs/piece_spec.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  describe '#valid_vector?' do
+    let(:forward_diagonal_vector) { Vector[2, 2] }
+    let(:forward_horizontal_vector) { Vector[5, 0] }
+    let(:knight_movement_vector) { Vector[2, 1] }
+    let(:backwards_pawn_move) { Vector[0, -1] }
+
+    shared_examples 'diagonal movement' do |valid|
+      it 'addresses diagonal movement' do
+        expect(subject.valid_vector?(forward_diagonal_vector)).to be valid
```
Comment (jacquesporveau) [inline spec/piece_specs/piece_spec.rb]: You are getting a little bit too close to another established naming convention with this test.

http://www.rubydoc.info/gems/rspec-rails/RSpec/Rails/Matchers#be_valid-instance_method

Comment (swcraig) [inline spec/piece_specs/piece_spec.rb]: @jacquesporveau Do you only dislike my use of `valid` as the parameter name? Or the entire structure of these shared_examples?

Comment (jacquesporveau) [inline spec/piece_specs/piece_spec.rb]: There is a matcher called `be_valid`. You are too close to that method. People could become confused and think you are using that at a quick glance.

**Comment by swcraig** [comment]
@jacquesporveau Does this look good to go?

File: spec/piece_specs/piece_spec.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+      it 'can move like a knight' do
+        expect(subject.valid_vector?(knight_movement_vector)).to be true
+      end
+    end
+
+    context 'the piece is a Pawn' do
+      subject { build(:piece, piece_type: Pawn, player: player) }
+
+      it_behaves_like 'can move diagonally', true
```
Comment (jacquesporveau) [inline spec/piece_specs/piece_spec.rb]: Can a Pawn move horizontally?

File: spec/piece_specs/piece_spec.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
+describe Piece do
+  let(:player) { build(:player) }
+
+  describe '#valid_vector?' do
+    let(:forward_diagonal_vector) { Vector[2, 2] }
+    let(:forward_horizontal_vector) { Vector[0, 5] }
+    let(:knight_movement_vector) { Vector[2, 1] }
+    let(:backwards_pawn_move) { Vector[0, -1] }
+
+    shared_examples 'can move diagonally' do |able_to_move_diagonally|
```
Comment (jacquesporveau) [inline spec/piece_specs/piece_spec.rb]: `it_behaves_like 'can move diagonally'` Doesn't ready very well. This is a tricky one to word but try to spend a little more time on this. If you can't come up with anything that's fine but let's see if we can.

**Comment by swcraig** [comment]
Thank you all!

**Pull request merged**